test/web/RawSocketHost.cs

```diff
@@ -0,0 +1,184 @@
+using Tmds.LinuxAsync.Transport;
+using Socket = Tmds.LinuxAsync.Socket;
+using SocketAsyncEventArgs = Tmds.LinuxAsync.SocketAsyncEventArgs;
+
+namespace web
+{
+    public class RawSocketHost
+    {
+        private const int BufferSize = 512;
+        private const string Response =
+            "HTTP/1.1 200 OK\r\nDate: Tue, 31 Mar 2020 14:49:06 GMT\r\nContent-Type: application/json\r\nServer: Kestrel\r\nContent-Length: 27\r\n\r\n{\"message\":\"Hello, World!\"}";
+
+        private static ReadOnlySpan<byte> RequestEnd => new byte[] {13, 10, 13, 10}; // "\r\n\r\n"
+        
+        private readonly string[] _args;
+        private readonly CommandLineOptions _options;
+
+        private IPEndPoint _serverEndpoint;
+
+        public RawSocketHost(CommandLineOptions options, string[] args)
+        {
+            _options = options;
+            _args = args;
+            
+            ConfigurationBuilder bld = new ConfigurationBuilder();
+            bld.AddCommandLine(_args);
+            var cfg = bld.Build();
+            string url = cfg["urls"];
+
+            string[] data = url.Substring(7).Split(':');
+            IPAddress ip = IPAddress.Parse(data[0]);
+            int port = int.Parse(data[1]);
+
+            _serverEndpoint = new IPEndPoint(ip, port);
+        }
+
+        public void Run()
+        {
+            RunAsync().GetAwaiter().GetResult();
+        }
+
+        public async Task RunAsync()
+        {
+            using Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            listener.Bind(_serverEndpoint);
+            listener.Listen(1024);
+            Console.WriteLine($"Raw server listening on {_serverEndpoint}");
+
+            bool deferSends = _options.DeferSends == true;
+            bool deferReceives = _options.DeferReceives == true;
+            bool runContinuationsAsynchronously =
+                _options.SocketContinuationScheduler == SocketContinuationScheduler.ThreadPool;
+
+            using ClientConnectionHandler clientHandler = 
+                new ClientConnectionHandler(deferSends, deferReceives, runContinuationsAsynchronously);
+            
+            while (true)
+            {
+                using Socket handlerSocket = await listener.AcceptAsync();
+
+                clientHandler.HandleClient(handlerSocket);
+            }
+        }
+
+        class ClientConnectionHandler : IDisposable
+        {
+            private Socket _socket;
+            private readonly byte[] _sendBuffer = Encoding.ASCII.GetBytes(Response);
+            private readonly byte[] _receiveBuffer = new byte[BufferSize];
+
+            private readonly SocketAsyncEventArgs _receiveArgs;
+            private readonly SocketAsyncEventArgs _sendArgs;
+            private readonly ManualResetEventSlim _sendEvent = new ManualResetEventSlim(true);
+            private readonly ManualResetEventSlim _receiveEvent = new ManualResetEventSlim();
+            
```

Comment:
We don't need the `ManualResetEventSlim`.
We should be able to do something like:

```cs
ClientConnectionHandler()
{
      _sendArgs.Completed += (s, a) => ReceiveRequest();
      _receiveArgs.Completed += (s, a) => OnReceive();
}

HandleClient() => ReceiveRequest()

ReceiveRequest()
{
	try
	{
                _receiveArgs.MemoryBuffer = _receiveBuffer.AsSpan(_bytesReceived);
		if (!_socket.ReceiveAsync(_receiveArgs))
		{
		    OnReceive();
		}
	}
	catch
	{
		Cleanup();
	}
}

OnReceive()
{
	int bytesTransferred = _receiveArgs.BytesTransferred;
	SocketError error = _receiveArgs.SocketError;

	if (error == SocketError.Success && count > 0)
	{
		_bytesReceived += bytesTransferred;
		if (RequestComplete(...))
		{
			_bytesReceived = 0;
			SendResponse();
		}
	}
	else
	{
		Cleanup();
	}
}

SendResponse()
{
	try
	{
		if (!_socket.SendAsync(_sendArgs))
		{
		    ReceiveRequest();
		}
	}
	catch
	{
		Cleanup();
	}	
}

Cleanup()
{
	_socket.Dispose();
}
```